transaction/std/src/memo_builder/gift_code_sender_memo_builder.rs

```diff
@@ -0,0 +1,276 @@
+
+        // Verify memo data
+        let blank_note = "";
+        let derived_note = MemoDecoder::decode_sender_note(memo_payload.get_memo_data());
+        assert_eq!(blank_note, derived_note);
+
+        // Create another memo builder
+        let mut builder_2 = GiftCodeSenderMemoBuilder::default();
+
+        // Set the cancellation index and then set another note on top of it
+        let note_2 = "Is anything actually real?";
+        builder_2.set_gift_code_sender_note(note).unwrap();
+        builder_2.set_gift_code_sender_note(note_2).unwrap();
+
+        // Build the memo payload and get the data
+        let amount_2 = Amount::new(666, 0.into());
+        let memo_payload_2 = build_change_memo_with_amount(&mut builder_2, amount_2).unwrap();
+        let derived_note_2 = MemoDecoder::decode_sender_note(memo_payload_2.get_memo_data());
+        assert_eq!(note_2, derived_note_2);
+    }
+
+    #[test]
+    fn test_gift_code_sender_memo_writes_unused_if_change_disabled() {
+        // Create memo builder
+        let mut builder = GiftCodeSenderMemoBuilder::default();
+
+        // Set the note
+        let note = "It's MEMO TIME!!";
+        builder.set_gift_code_sender_note(note).unwrap();
+
+        // Disable change memos
+        builder.disable_gift_code_change_memo();
+
+        // Build the memo payload
+        let amount = Amount::new(42, 0.into());
+        let memo_payload = build_change_memo_with_amount(&mut builder, amount).unwrap();
+        assert_eq!(memo_payload.get_memo_data(), &[0u8; 64]);
+    }
+
+    #[test]
+    fn test_gift_code_sender_note_builder_fails_after_attempting_invalid_note() {
+        // Create memo builder
+        let mut builder = GiftCodeSenderMemoBuilder::default();
+
+        // Attempt to set an invalid note
+        let note_bytes = [b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN + 1];
+        let note = std::str::from_utf8(&note_bytes).unwrap();
+        let result = builder.set_gift_code_sender_note(note);
+        assert!(matches!(result, Err(NewMemoError::BadInputs(_))));
+
+        // Fail to build memo after failing to set valid note
+        let amount = Amount::new(42, 0.into());
+        let memo_payload = build_change_memo_with_amount(&mut builder, amount);
+        assert!(matches!(memo_payload, Err(NewMemoError::BadInputs(_))));
+    }
+
+    #[test]
+    fn test_gift_code_sender_note_builder_succeeds_after_resetting_invalid_note_with_valid_note() {
+        // Create memo builder
+        let mut builder = GiftCodeSenderMemoBuilder::default();
+
+        // Attempt to set an invalid note
+        let note_bytes = [b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN + 1];
+        let note = std::str::from_utf8(&note_bytes).unwrap();
+        let result = builder.set_gift_code_sender_note(note);
+        assert!(matches!(result, Err(NewMemoError::BadInputs(_))));
+
+        // Reset with valid note
+        let valid_note = "Im within the byte length, but validity is a subjective concept";
```

Comment:
Should check boundary case of note at maximum length:
```suggestion
        let valid_note_bytes = [b'6'; GiftCodeSenderMemo::MEMO_DATA_LEN];
        let valid_note = std::str::from_utf8(&valid_note_bytes).unwrap();
```

Reply:
Also probably better!